src/lib.rs

```diff
@@ -187,66 +183,85 @@ impl State {
+            if let Err(e) = second_user.add_friendship(&friendship).await {
+                // cannot use this friendship since user couldn't find a valid room
+                log::info!("Couldn't reuse friendship: {}", e);
+                continue;
+            }
+            progress_bar.inc(1);
+            progress_bar.tick();
+            self.friendships.push(friendship);
+        }
 
-                friendship
-            } else {
-                let (first_user, second_user) = self.get_random_friendship();
-                let friendship = Friendship::from_users(first_user, second_user);
+        let friendships_to_create = amount_of_friendships - self.friendships.len();
 
-                futures.push(join_users_to_room(
-                    first_user,
-                    second_user,
-                    friendship.clone(),
-                    &progress_bar,
-                    self.config.room_creation_retry_attempts,
-                    self.config.room_creation_max_resource_wait_attempts,
-                ));
+        log::info!(
+            "Reused friendships {}, will try to create {} new friendships",
+            self.friendships.len(),
+            friendships_to_create
+        );
 
-                friendship
-            };
+        // create remaining friendships
+        let mut futures = vec![];
+        for _ in 0..amount_of_friendships - self.friendships.len() {
+            let (first_user, second_user) = self.get_random_friendship();
+            let friendship = Friendship::from_users(first_user, second_user);
+
+            futures.push(join_users_to_room(
+                first_user,
+                second_user,
+                friendship.clone(),
+                self.config.room_creation_retry_attempts,
+                self.config.room_creation_max_resource_wait_attempts,
+                self.config.update_account_data,
+            ));
 
-            self.friendships.push(friendship);
+            log::info!("Creating friendship: {:?}", friendship);
         }
-
         let stream_iter = iter(futures);
         let mut buffered_iter = stream_iter.buffer_unordered(self.config.room_creation_throughput);
 
         while let Some(res) = buffered_iter.next().await {
-            if let Some((user1, user2)) = res {
-                self.users_state.add_friendship(user1, user2)
+            if let Some((friendship, user1, user2)) = res {
+                // friendship created
+                progress_bar.inc(1);
+                self.users_state.add_friendship(user1, user2);
+                self.friendships.push(friendship);
+            } else {
+                log::info!("Friendship was not created :(");
             }
         }
 
-        self.friendships.sort();
+        let missing_friendships = amount_of_friendships - self.friendships.len();
+        if missing_friendships > 0 {
+            log::info!("Couldn't create {missing_friendships} friendships");
```

Comment:
I think we should `panic!` in this case.

Reply:
yes, also we can retry a few times before `panic!`

anyway, it's possible that one step has missing friendships but the next to be complete.